### PR TITLE
Store API Comments

### PR DESCRIPTION
**Pull request by kavishkafernando**
### **Purpose**
Add comments feature to WSO2 APIM-3.0. 

### **Goals**
Display the comments (which are retrieved from the back-end) in API Store Comments Section.
Provide the facilities to add and delete comments in API Store.

Screen Shot 2019-08-30 at 10 38 20 AM

### **Adding a Comment**
Screen Shot 2019-08-30 at 5 41 29 PM

### **Display Comments**
Screen Shot 2019-08-30 at 10 38 20 AM

### **Delete Comment**

-  The comment author and the admin can delete comments.

Screen Shot 2019-08-30 at 5 42 39 PM

### **Overview**
The 3 latest comments are displayed in the Overview page of the API.
Screen Shot 2019-08-30 at 5 44 27 PM

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml [hunk 1]
```diff
@@ -950,6 +952,12 @@ paths:
             Comments list is returned.
           schema:
             $ref: '#/definitions/CommentList'
+        401:
+          description: |
+            Un authorized.
+            The user is not authorized to view the application .
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml]: Un authorized =>  Unauthorized
Shall we fix the description as well?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml [hunk 2]
```diff
@@ -999,6 +1008,12 @@ paths:
             Invalid request or validation error.
           schema:
             $ref: '#/definitions/Error'
+        401:
+          description: |
+            Un authorized.
+            The user is not authorized to view the application .
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml]: Same applies here.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java [hunk 3]
```diff
@@ -167,33 +169,124 @@ public Response apisApiIdGraphqlSchemaGet(String apiId, String ifNoneMatch, Stri
...
+            APIIdentifier apiIdentifier = APIMappingUtil.getAPIIdentifierFromUUID(apiId, requestedTenantDomain);
+
+            Comment comment = CommentMappingUtil.fromDTOToComment(body, username, apiId);
+            int createdCommentId = apiConsumer.addComment(apiIdentifier, comment, username);
+            Comment createdComment = apiConsumer.getComment(apiIdentifier, createdCommentId);
+            CommentDTO commentDTO = CommentMappingUtil.fromCommentToDTO(createdComment);
+
+            String uriString = RestApiConstants.RESOURCE_PATH_APIS + "/" + apiId +
+                    RestApiConstants.RESOURCE_PATH_COMMENTS + "/" + createdCommentId;
+            URI comments = new URI(uriString);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java]: Shall we rename URI `comments` to `uri`?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java [hunk 4]
```diff
@@ -167,33 +169,124 @@ public Response apisApiIdGraphqlSchemaGet(String apiId, String ifNoneMatch, Stri
...
-        // do some magic!
-        return Response.ok().entity("magic!").build();
+    public Response getAllCommentsOfAPI(String apiId, String xWSO2Tenant, Integer limit, Integer offset,
+                                        MessageContext messageContext) {
+        String requestedTenantDomain = RestApiUtil.getRequestedTenantDomain(xWSO2Tenant);
+        try {
+            APIConsumer apiConsumer = RestApiUtil.getLoggedInUserConsumer();
+            APIIdentifier apiIdentifier = APIMappingUtil.getAPIIdentifierFromUUID(apiId, requestedTenantDomain);
+
+            Comment[] createdComment = apiConsumer.getComments(apiIdentifier);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java]: Shall we rename `createdComment` with a proper one?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/source/src/app/components/Apis/Details/Comments/Comment.jsx [hunk 5]
```diff
@@ -190,36 +191,37 @@ class Comment extends React.Component {
...
-                    Alert.error(
-                        intl.formatMessage({
-                            defaultMessage: 'Something went wrong while deleting comment',
-                            id: 'Apis.Details.Comments.Comment.something.went.wrong',
-                        })
-                            + ' - '
-                            + commentIdOfCommentToDelete,
-                    );
-                }
+                // if (error.response) {
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/source/src/app/components/Apis/Details/Comments/Comment.jsx]: Shouldn't we keep these Alerts to identify failures?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml (hunk 1 above)
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml (hunk 2 above)
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java (hunk 3 above)
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java (hunk 4 above)
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/source/src/app/components/Apis/Details/Comments/Comment.jsx (hunk 5 above)
Comment (kavishkafernando) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/source/src/app/components/Apis/Details/Comments/Comment.jsx]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml (hunk 1 above)
Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml]: "The user is not authorized to view the application:" description may need to be fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml (hunk 2 above)
Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml]: Same here

**Pull request merged**